### ISSUE TITLE
Discussion on applying the MVC pattern

### ISSUE THREAD

**Issue by tian-chaiyaporn** [tian-chaiyaporn/thinkful-capstone-1#2]
@joelwallis
So here is a basic plan on which module goes where:
Model
  - HousePriceModel
     - RequestAPI from Quandl (private sub-module)
     - ProcessData from Quandl and store in cache/local storage (private sub-module)
  - AreaCodeModel

Controller
  - App.js (this acts as the mediator) - I think the routing system is to be set up here? Or should I try to separate it into another Controller module?
  -Utils.js (global utility functions - for calculating percentage change, formatting strings etc...)

View
  - Map (initialize google map)
  - Markers (create Markers as the data is passed on from App.js)
  - Notify (links to App.js, and gets called whenever there is no data or when users have to wait)

What do you think about this arrangement? Should I go ahead and try to implement it in code?

**Comment by tian-chaiyaporn** [comment]
I visualized it here so it's a little easier to understand about which module should be M V or C

![mvc](https://cloud.githubusercontent.com/assets/20303596/26091214/e59de9c0-39d7-11e7-962e-8c88533529b7.png)

**Comment by joeljuca** [comment]
After trying to use MVC as our architecture design we decided to simplify it by using a simpler variation of MVP. We'll have three moments in our app:

- **Rendering**  
  Renders the app from the current state - which serves as our source of truth
- **User actions**  
  Actions made by users (clicking around, toggling and/or choosing options, etc.) firing events that might change the state
- **State changes**  
  Changes to the app state, that by its time fires the rendering process, updating the page according to the state

Below is a enterprise grade explanation of what I mean:

![8123457890234](https://cloud.githubusercontent.com/assets/673884/26422841/cd36659e-40a1-11e7-850b-882f74a45486.png)

**Comment by tian-chaiyaporn** [comment]
Hey @joelwallis , so after our discussion, I've come up with a new design for MVP, most functionalities stay the same, but the coupling and application flow changes a little. Let me know what you think:
![mvp](https://cloud.githubusercontent.com/assets/20303596/26478510/ae877cc4-419a-11e7-9bfc-b6e54e4d9e0d.png)

**Comment by tian-chaiyaporn** [comment]
@joelwallis Here is the new architecture revision after our discussion. I will create a new github repo to do a basic version of this and add you as collaborator :)


![mvp-revised](https://cloud.githubusercontent.com/assets/20303596/26514962/dfa76202-4243-11e7-8cdd-67ab84f6d62d.png)

**Comment by joeljuca** [comment]
@tian-chaiyaporn awesome! That one really seems to be easier to understand.

Let me know when you create the repo for the project that will validate that architecture - and congrats for being such a dedicated student! I'm quite proud of you! 😍 haha

**Comment by tian-chaiyaporn** [comment]
@joelwallis after a few trials and errors, I decided to change the app architecture to this one instead. This is because the UI is quite complex, so I can't connect the view straight to the model. I also had to extract a few more modules because they were becoming too large.

I ended up not using the observer abstraction because there were too many event types that it was taking me too much time to handle them. I might add this in later when I have more time after the capstone. However, I think the modules are still sufficiently decoupled from each other.

![drawing-3-custommvp](https://cloud.githubusercontent.com/assets/20303596/26785107/d2af397a-49ce-11e7-9f2d-d74f91ad4a0d.png)
